### ISSUE TITLE
Design compile-time analysis of type hierarchies

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#2583]
Part of #2544.

The weaver (and code host loader) should support compile-time analysis of database class hierarchies, in a way that can identify matrices needed to construct underying type system, as defined in #2544 and in [this specification](https://github.com/Starcounter/Starcounter/wiki/Under-the-hood;-the-VMDBMS-dynamic-type-system).

From any application, a set of hierarchies should be established, based on leaf classes. At the end, mapping between instance class hierarchies and type class hierarchies should be made, and all defined compile-time constraints should be possible to evaluate.

**Comment by per-samuelsson** [comment]
I had an implementation yesterday that I thought was quite close to completion, but now I'm struggling to redesign it in some way, since it was not sufficient for models where the type hierarchy is multilevel in all directions, and instance classes puts a reference right into the middle of it, like this:

```C#
[Database] public class VehicleModel {}
[Database] public class CarModel : VehicleModel {}
[Database] public class SUVType {}
[Database] public class StandardSUVKind {}
[Database] public class StandardSUV: SUVType { [Type] public StandardSUVKind Kind; }
[Database] public class SUV : CarModel { [Type] public SUVType Type; }
[Database] public class Combi : CarModel {}

// The single instance class
[Database] public class Car { [Type] public CarModel Model; }
```

The facit matrix is solvable, but an implementation that can verify its correctness and produce a logical model of it is what I'm struggling with.

**Comment by per-samuelsson** [comment]
About this model from above,

>```C#
class VehicleModel {}
class CarModel : VehicleModel {}
class SUVType {}
class StandardSUVKind {}
class StandardSUV: SUVType { [Type] StandardSUVKind Kind; }
class SUV : CarModel { [Type] SUVType Type; }
class Combi : CarModel {}
// The single instance class
class Car { [Type] CarModel Model; }
```

 the facit matrix is

Car | Type | Type |Type | Type|Type
-----|-----|-----|------|---|---|---
 |100|VehicleModel|101|102|103
 |Car|CarModel|200|201|202
 |300,301|SUV,Combi|SuvType,302|303,304|304,306
 |400,401|402,403|StandardSuv,405|StandardSUVKind,406|407,408

where numbers indicate unique type objects with no name (i.e. "auto" types), and named types indicate **name**, not host class.

The "broadest path" from Car instance class is Car->SUV->StandardSUV->StandardSuvKind->auto.

**Comment by per-samuelsson** [comment]
Another little model I'm giving the algorithm as input to see it works: 

```C#
class HatModel { [Type] HatModelForm Model; }
class HatModelForm { }
class HighHat : HatModelForm { }
class RoundHat : HatModelForm { }
class Hat { [Type] HatModel Model; }
```

Hat | Type | Type |Type | Type
-----|-----|-----|------|---|---
 |Hat|HatModel|HatModelForm|100
 |200,201|202,203|HighHat, RoundHat|204,205

**Comment by per-samuelsson** [comment]
Here's one I don't think we have covered before.

```C#
class VehicleModel {}
class CarModel : VehicleModel {}
class MazdaModel : CarModel {}

class Car {[Type] VehicleModel Model;}
class Opel : Car {}
```

I guess it must produce this:

Instance | Type | Type |Type
-----|-----|-----|------|---
 |Car|VehicleModel|100
 |Opel|CarModel|200
 |300|MazdaModel|301

**Comment by k-rus** [comment]
I have the same understanding of what will be produced.

**Comment by per-samuelsson** [comment]
The following model (an submodel from [here](https://github.com/Starcounter/Starcounter/wiki/Under-the-hood;-the-VMDBMS-dynamic-type-system#1100-complex-milestone))

```C#
class CarModel {}
class EuropeanModel : CarModel {}
class SwedishModel : EuropeanModel {}

class Car { [Type] CarModel Model; }
class EuropeanCar : Car { [Type] EuropeanModel Model; }
class ScandinavianCar : EuropeanCar {}
class SwedishCar : ScandinavianCar { [Type] SwedishModel Model; }
```

 we resolve like the below I assume?

Instance | Type | Type |Type
-----|-----|-----|------|---
 |Car|CarModel|100
 |EuropeanCar|EuropeanModel|200
 |ScandinavianCar|300|301
 |SwedishCar|SwedishModel|400

**Comment by k-rus** [comment]
> we resolve like the below I assume?

I think the same

**Comment by per-samuelsson** [comment]
This has been reevaluated, and will be put in the backlog. Part of the functionality implemented has been cherrypicked and merged as part of #2630.
